File: behavioral/chain-of-responsibilities-example.py
class BuyIngredients:

    @staticmethod
    def run(input_data):
        # do something with input_data and return the result,
        # go to the market and buy whatever is in input data
        process_result = f"I have {input_data['money']} money and just bought {str(input_data['ingredients'])}"
        return process_result


class Deliver:
    @staticmethod
    def run(input_data):
        # do something with input_data and return the result,
        # get a cab and deliver it to restaurant
        process_result = input_data
        process_result += '\nI have to get a cab and deliver it to restaurant'
        return process_result


class Cook:

    def __init__(self, name):
        self.__name = name

    def run(self, input_data):
        # do something with input_data and return the result,
        # ask the cook to cook the dish
        process_result = input_data
        process_result += f'\nIngredients are delivered, I ({self.__name}) cooked the dish and it is ready. '
        return process_result


class OrderSalad:

    @staticmethod
    def run(input_data):
        process_result = input_data
        process_result += f'\nOrdering and adding salad'
        return process_result


class Chain:
    def __init__(self):
        self.__process = []

    def add_process(self, p):
        self.__process.append(p)

    def run(self, init_data):
        step_data = init_data
        for p in self.__process:
            step_data = p.run(step_data)
        return step_data

    def clear(self):
        self.__process = []


if __name__ == "__main__":
    data = {'money': 100, 'ingredients': {'flower': 100, 'raisins': 50, 'salt': 5}}

    chain = Chain()
    chain.add_process(BuyIngredients())
    chain.add_process(Deliver())
    chain.add_process(Cook('kamran'))
    result = chain.run(data)
    print(result)

    print('--------------------')

    chain.clear()
    chain.add_process(BuyIngredients())
    chain.add_process(Deliver())
    chain.add_process(Cook('tina'))
    chain.add_process(OrderSalad())
    result = chain.run(data)
    print(result)

File: behavioral/chain-of-responsibilities-example.js
class BuyIngredients {
    static run(inputData) {
        const processResult = `I have ${inputData.money} money and just bought ${JSON.stringify(inputData.ingredients)}`;
        return processResult;
    }
}

// ------------------------------------------------------------------
class Deliver {
    static run(inputData) {
        let processResult = inputData;
        processResult += '\n I have to get a cab and deliver it to restaurant.';
        return processResult;
    }
}

// ------------------------------------------------------------------
class Cook {
    constructor(name) {
        this.name = name;
    }

    run(inputData) {
        let processResult = inputData;
        processResult += `\n Ingredients are delivered, I (${this.name}) cooked the dish and it is ready.`
        return processResult;
    }
}

// ------------------------------------------------------------------
class OrderSalad {
    static run(inputData) {
        let processResult = inputData;
        processResult += `\n Adding salad to my order.`
        return processResult;
    }
}

// ------------------------------------------------------------------
class Chain {
    constructor() {
        this.process = [];
    }

    addProcess(p) {
        this.process.push(p);
    }

    run(init_data) {
        let stepData = init_data;
        for (const p of this.process) {
            stepData = p.run(stepData);
        }
        return stepData;
    }

    clear() {
        this.process = [];
    }
}

// ------------------------------------------------------------------
const data = {'money': 100, 'ingredients': {'flour': 100, 'raisins': 50, 'salt': 5}};

const chain = new Chain();
chain.addProcess(BuyIngredients);
chain.addProcess(new Cook('Tina'));
chain.addProcess(Deliver);

const result = chain.run(data);
console.log(result);

console.log();
console.log('--------------------');
console.log();

chain.clear();
chain.addProcess(BuyIngredients);
chain.addProcess(new Cook('Kamran'));
chain.addProcess(Deliver);

chain.addProcess(OrderSalad);
const result1 = chain.run(data);
console.log(result1);